### PR TITLE
Add throttling to error e-mails

### PR DESCRIPTION
**Pull request by dmgaldi**
## Overview
Added throttling for error e-mails.

Things I'm unsure about:
* Should throttling config be moved to model config?
* E-mail contents and subject line. Maybe they should be more consistent with error emails?

## Testing
* Set throttling on my dev site to burst size of 1 and verified a single error e-mail came through.

### CODE REVIEW

**Review by ryanrdoherty — CHANGES_REQUESTED** [review]
Added that one comment + our discussion about converting to a non-static class.

**Review by ryanrdoherty — CHANGES_REQUESTED** [review]

File: Model/src/main/java/org/eupathdb/common/errors/ErrorEmailThrottler.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Wraps {@link TokenBucketPermitDistributor}, keeping track of how many times an e-mail was throttled. Sends an e-mail
+ * once per hour summarizing throttles.
+ */
+public class ErrorEmailThrottler {
+  private static final Logger LOG = Logger.getLogger(ErrorEmailThrottler.class);
+  private static final String WDK_PROP_EMAIL_BURST_THROTTLE_SIZE_KEY = "EMAIL_BURST_THROTTLE_SIZE";
+  private static final String WDK_PROP_EMAIL_SUSTAINED_THROTTLE_KEY= "EMAIL_SUSTAINED_THROTTLE_RATE_PER_SECOND";
```
[the remark this answers is not in the document]
Comment (dmgaldi) [reply Model/src/main/java/org/eupathdb/common/errors/ErrorEmailThrottler.java:31]: That's correct. We want to fill the bucket once every 10 minutes. Would you prefer the units to be specified as per minute or hour and the converted when passed to the underlying rate limiter?

File: Model/src/main/java/org/eupathdb/common/errors/ErrorEmailThrottler.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  public ErrorEmailThrottler(WdkModel wdkModel) {
+    double burstSize = Optional.ofNullable(wdkModel.getProperties().get(WDK_PROP_EMAIL_BURST_THROTTLE_SIZE_KEY))
+        .map(Double::parseDouble)
+        .orElse(DEFAULT_EMAIL_BURST_THROTTLE_SIZE);
+    double sustainedRate = Optional.ofNullable(wdkModel.getProperties().get(WDK_PROP_EMAIL_SUSTAINED_THROTTLE_KEY))
+        .map(Double::parseDouble)
+        .orElse(DEFAULT_EMAIL_SUSTAINED_THROTTLE_RATE);
+    permitDistributor = new TokenBucketPermitDistributor(sustainedRate, burstSize);
+    scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    scheduledExecutorService.scheduleAtFixedRate(this::sendThrottlingReport, 0L, 1, TimeUnit.HOURS);
```
[the remark this answers is not in the document]
Comment (dmgaldi) [reply Model/src/main/java/org/eupathdb/common/errors/ErrorEmailThrottler.java:50]: Theoretically, yes when tomcat is shutdown we should gracefully close this. Where would this belong on the site lifecycle? In `EuPathSiteSetup` called by `ApplicationInitListener`?

File: Model/src/main/java/org/eupathdb/common/errors/ErrorEmailThrottler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryanrdoherty) [reply Model/src/main/java/org/eupathdb/common/errors/ErrorEmailThrottler.java:31]: So the units added in the "rate per second" is not "permits" but "burst size permits" i.e. "fills"?  If this is correct, then the actual fill rate depends on both the burst size and the throttle rate?  This seems more confusing to me than simply using "permits" as the units for both and having them be independent (total size and # added per second).  Maybe I'm still just thinking about this all wrong?

File: Model/src/main/java/org/eupathdb/common/errors/ErrorEmailThrottler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanrdoherty) [reply Model/src/main/java/org/eupathdb/common/errors/ErrorEmailThrottler.java:50]: Let's meet about this.  Thinking of two ways to do it and we can decide together which one is better.

**Pull request merged**